uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-next-leave.component.js

```diff
@@ -0,0 +1,178 @@
+    }]
+  });
+
+  nextLeaveController.$inject = ['$scope', 'LeaveRequest', 'OptionGroup',
+    'shared-settings'];
+
+  function nextLeaveController ($scope, LeaveRequest, OptionGroup,
+  sharedSettings) {
+    var childComponentName = 'leave-widget-next-leave';
+    var dayTypes = [];
+    var vm = this;
+    var statusColoursMap = {
+      'admin_approved': 'success',
+      'approved': 'success',
+      'awaiting_approval': 'warning',
+      'more_information_required': 'primary'
+    };
+
+    vm.balanceDeduction = 0;
+    vm.nextLeaveRequest = null;
+    vm.requestStatus = {};
+
+    vm.$onChanges = $onChanges;
+
+    /**
+     * Initializes the controller by emiting a child is loading event.
+     */
+    (function init () {
+      $scope.$emit('LeaveWidget::childIsLoading', childComponentName);
+    })();
+
+    /**
+     * Implements $onChanges method for AngularJS Components. Waits for bindings
+     * to be ready, and then loads the next leave requests. When it's done, it
+     * emits a child is ready event.
+     */
+    function $onChanges () {
+      if (bindingsAreReady()) {
+        loadNextLeaveRequest().finally(function () {
+          $scope.$emit('LeaveWidget::childIsReady', childComponentName);
+        });
+      }
+    }
+
+    /**
+     * Returns true when contact id and leave request statuses bindings are
+     * ready.
+     *
+     * @return {Boolean}
+     */
+    function bindingsAreReady () {
+      return vm.contactId && vm.leaveRequestStatuses &&
+        vm.leaveRequestStatuses.length;
+    }
+
+    /**
+     * Returns the label for the day type id provided.
+     *
+     * @param {String} dayTypeId - the id for the day type.
+     * @return {String}
+     */
+    function getDayTypeLabel (dayTypeId) {
+      var dayType = _.find(dayTypes, function (dayType) {
+        return +dayType.value === +dayTypeId;
+      });
+
+      return dayType.label;
+    }
+
+    /**
+     * Returns the status for the next leave request.
+     *
+     * @return {Object}
+     */
+    function getNextLeaveRequestStatus () {
+      return _.find(vm.leaveRequestStatuses, function (status) {
+        return +status.value === +vm.nextLeaveRequest.status_id;
+      });
+    }
+
+    /**
+     * Returns a list of status ids.
+     *
+     * @return {Array}
+     */
+    function getStatusIds () {
+      return vm.leaveRequestStatuses.map(function (status) {
+        return status.value;
+      });
+    }
+
+    /**
+     * Loads and stores all the possible day types for leave requests.
+     *
+     * @return {Promise}
+     */
+    function loadDayTypes () {
+      return OptionGroup.valuesOf('hrleaveandabsences_leave_request_day_type')
+        .then(function (_dayTypes_) {
+          dayTypes = _dayTypes_;
+        });
+    }
+
+    /**
+     * Loads the next leave request for the contact that has been approved or
+     * is under review.
+     *
+     * @return {Promise}
+     */
+    function loadNextLeaveRequest () {
```

Comment:
This is another example of a function that does too many things

1. It loads the next leave request
2. It assigns day labels to it (which in turns fetches the day types from the api!)
3. It updates the balance deduction
4. it updates the request status

It's obvious that this function has
1. A very misleading name
2. Does too many things as said before

The code should be refactored in something like this

```js
function $onChanges () {
  if (!bindingsAreReady()) { // avoids nesting
    return;
  }

  $q.all([
    loadDayTypes(),
    loadNextLeaveRequest()
  ])
  .then(function () {
    if (vm.nextLeaveRequest) {
      processNextLeaveRequestData();
    }
  })
  .finally(function () {
    $scope.$emit('LeaveWidget::childIsReady', childComponentName);
  });
}

function processNextLeaveRequestData () {
  mapDateTypeLabels();
  updateBalanceDeduction();
  updateRequestStatus();
}
```